docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,215 @@
+  repeated Event events     = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace  = 8;
++ int64          priority   = 9;
++ string         sender     = 10;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer and fee payer 
+of the transaction, the `priority` could be the cumulative sum of the fee(s).
+
+Only `sender` is required, while `priority` can be omitted which would result in
+using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:*Tx>`, i.e a mapping from `sender` to a single `*Tx`
+object, where each `*Tx` is the next valid and processable transaction from the
+given `sender`.
+
+On top of this mapping, we index all transactions by priority using a thread-safe
+priority queue, i.e. a [max heap](https://en.wikipedia.org/wiki/Min-max_heap).
+When a proposer is ready to select transactions for the next block proposal,
+transactions are selected from this priority index by highest priority order.
+When a transaction is selected and reaped, it is removed from this index and
+from the `<sender:*Tx>` mapping.
+
+We define `Tx` as the following data structure:
+
+```go
+type Tx struct {
+  // Tx represents the raw binary transaction data.
+  Tx []byte
+
+  // Priority defines the transaction's priority as specified by the application
+  // in the ResponseCheckTx response.
+  Priority int64
+
+  // Sender defines the transaction's sender as specified by the application in
+  // the ResponseCheckTx response.
+  Sender string
+
+  // Index defines the current index in the priority queue index. Note, if
+  // multiple Tx indexes are needed, this field will be removed and each Tx
+  // index will have its own wrapped Tx type.
+  Index int
+}
+```
+
+### Eviction
+
+Upon successfully executing `CheckTx` for a new `Tx` and the mempool is currently
+full, we must check if there exists a `Tx` of lower priority that can be evicted
+to make room for the new `Tx` with higher priority and with sufficient size
+capacity left.
+
+If such a `Tx` exists, we find it by obtaining a read lock and sorting the
+priority queue index. Once sorted, we find the first `Tx` with lower priority and
+size such that the new `Tx` would fit within the mempool's size limit. We then
+remove this `Tx` from the priority queue index as well as the `<sender:*Tx>`
+mapping.
+
+This will require additional `O(n)` space and `O(n*log(n))` runtime complexity.
+
+### Gossiping
+
+When gossiping new transactions to peers, we first obtain a read lock on the
+priority queue index. If we have not gossiped the top priority `Tx` for the
+given peer, we return this `Tx` as the next transaction to gossip. Otherwise,
+we sort the priority queue index and traverse in descending priority order until
```

Comment:
Is this not always sorted? Should it be?

Reply:
A priority queue, i.e. max-heap, is not sorted. We only need sorted iteration mainly when gossiping. We could consider using a sorted array index instead of a priority-queue index.

Main operations include:

- Adding to index
- Removing from index
- Sorted iteration on index